alternator/streams.cc

```diff
@@ -482,47 +476,130 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
         auto i = topologies.begin();
         auto e = topologies.end();
         auto prev = e;
-        auto threshold = db_clock::now() - confidence_interval(db);
         auto shards = rjson::empty_array();
 
         std::optional<shard_id> last;
 
+        // filter out cdc generations older than the table or now() - 24h
+        auto low_ts = std::max(as_timepoint(schema->id()), db_clock::now() - 24h);
+        while (i != e) {
+            auto j = i;
+            if (++j != e && j->first < low_ts) {
+                i = j;
+                continue;
+            }
+            // next is younger.
+            break;
+        }
+
+        auto first = i;
+
+        if (shard_start) {
+            i = std::find_if(i, e, [&](auto& p) {
+                return p.first == shard_start->start_time;
+            });
+        }
+
+        auto cmp = [](const cdc::stream_id& id1, const cdc::stream_id& id2) {
+            auto i1 = id1.index();
+            auto i2 = id2.index();
+            return i1 < i2 || (i1 == i2 && id1.token() < id2.token());
+        };
+
+        // need a prev even if we are skipping stuff
+        if (i != first) {
+            auto j = i;
+            --j;
+            std::sort(j->second.streams().begin(), j->second.streams().end(), cmp);
+            prev = j;
+        }
+
         for (; limit > 0 && i != e; prev = i, ++i) {
-            auto& [ts, sv] = *i;
+            auto [ts, sv] = *i;
 
             last = std::nullopt;
 
-            for (auto& id : sv.streams()) {
-                if (shard_start && shard_start->id != id) {
-                    continue;
+            // TODO: get rid of this once cdc generations
```

Comment:
Then why is there an "expired" field in the topology? It seems to me it would be more practical (if slightly redundant) if the generation management set/updated this table column when creating the new generation. To me, extracting it like this feels like a workaround that deserves the todo. 

The code below simply allows for the field to be set. If you want to change it to be a permanent solution, then the column should go.

Reply:
It's used for a generation garbage-collection mechanism that exists in my head but not in the code yet. The mechanism doesn't guarantee that `it->second.expired() == std::next(it)->first` because it doesn't need it and doesn't care.

`G.expired = X` means the following: if `now() - X > max{table.cdc.ttl for all tables with cdc enabled}`, then `G` can be safely removed (because no table can possibly have data in `G` anymore). In particular, if `G.expired = X` is OK, then  `G.expired = X + 10 days` is also OK.

The mechanisms works as follows:
- someone sees at some point that generation `G` is not operating anymore, i.e. there's a newer generation and everybody is writing to that new generation, and not to `G` anymore.
- he sets `G.expired` to `now()`. Note that whoever does it doesn't need to be a dedicated process, it can be done "on the occasion".
- someone else periodically queries `time, expired` from CDC description tables. (say, once a day)
- that someone checks the condition `now() - expired > max{table.cdc.ttl for all tables with cdc enabled}` for each entry queried. If the condition is true, he removes the entry.
Note that the condition may always remain false if there exists at least one table with `ttl = 0` (no data expiration). I guess it can be made weaker by also comparing table creation times, but still there might be a case where we simply cannot ever GC a generation.

Reply:
Fair enough, will kill it and always use next:s timestamp.